### PR TITLE
Feature: add functionality to check completed resources

### PR DESCRIPTION
**Pull request by ristofer**
# Base PR
- PR based on main.

# Context
- With this PR, we are adding the functionality to check resources as completed. It is the same functionality that the learning units has. The model and methods are based on the ones from the learning unit.

# Changelog
- Create a new table completed_resources that has a reference to a user and a resource.
- Create the corresponding model CompletedResource.
- Add new routes /api/resources/{resource_id}/completed.
- Add methods (GET) completed, (POST) complete_resource and (DELETE) uncomplete_resource on resources controller.
- Create new test for each of the new endpoints of completed resource functionality.
- Add creation of CompletedResources on the seed file.
- Update swagger.json file with new endpoints.
- Update annotation of learning_unit model.

# QA
- Run migrations (rails db:migrate)
- Run db reset (rails db:reset) to update the seeds
- Go to http://localhost:3001/api-docs/index.html 
- Execute GET, POST and DELETE on endpoint /api/resources/{resource_id}/completed.

### CODE REVIEW

File: app/controllers/api/resources_controller.rb [hunk 1]
```diff
@@ -52,18 +57,41 @@ def create
...
     end
 
-    private
+    def completed
+      resource_id = Resource.find(params[:resource_id])
+      completed_by_user = CompletedResource.find_by(
+        user: current_user,
+        resource_id:
+      )
+      completed = completed_by_user ? true : false
```
Comment (valentinjadot) [inline app/controllers/api/resources_controller.rb:66]: `completed = completed_by_user.present?`

File: app/controllers/api/resources_controller.rb [hunk 2]
```diff
@@ -52,18 +57,41 @@ def create
...
-        'Sin evaluación'
+        bad_request
       end
     end
 
+    def uncomplete_resource
+      completed_resource = CompletedResource.find_by(
+        resource_id: params[:resource_id], user_id: current_user
+      )
+      if completed_resource
```
Comment (valentinjadot) [inline app/controllers/api/resources_controller.rb:85]: `if completed_resource.present?`

File: app/services/resources/resource_evaluation_service.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
+module Resources
+  class ResourceEvaluationService
```
Comment (valentinjadot) [inline app/services/resources/resource_evaluation_service.rb:2]: ahi pasaria el `resource_id` al  momento de instanciar el servicio`def initialize(resource_id)`
onda → `Resources::ResourceEvaluationService.new(resource_id:).calculate_average_evaluation`


lo haria asi, avisame si quieres que te lo comento de vivo:


```
module Resources
  class ResourceEvaluationService
    def initialize(resource_id)
      @resource_id = resource_id
    end

    def calculate_average_evaluation
      average_evaluation.presence || 'Sin evaluación'
    end

    private

    def rounded_evaluation
      average_evaluation&.round(1)
    end

    def average_evaluation
      @average_evaluation ||= ResourceEvaluation.where(resource_id: @resource_id)
                                                .average(:evaluation)
    end
  end
end
```

y ahi puedes tener en el servicio un e

**Review by valentinjadot — CHANGES_REQUESTED** [review]
Dejé algunos comentarios para aprendisaje.

Creo que por la próxima, podrias separarlo en al menos 2 PRs (ej: uno con las migaciones y modelos, luego un otro con el controlador y servicio, etc.).

By the way, bacan que usas un servicio.

Podria ser también un comando si que se limita a entregar el average no más. 

Si piensas que va a hacer más, el servicio esta bien

File: app/controllers/api/resources_controller.rb (hunk 1 above)
Comment (ristofer) [reply app/controllers/api/resources_controller.rb]: Done :)

File: app/controllers/api/resources_controller.rb (hunk 2 above)
Comment (ristofer) [reply app/controllers/api/resources_controller.rb]: Done :)

File: app/services/resources/resource_evaluation_service.rb (hunk 3 above)
Comment (ristofer) [reply app/services/resources/resource_evaluation_service.rb:2]: Done :) Thanks for all the comments!

File: spec/requests/api/resources_controller_spec.rb [hunk 4]
```diff
@@ -207,4 +214,94 @@
...
+
+  path '/api/resources/{resource_id}/completed' do
+    get 'Return Resource completed status' do
+      tags 'Resources'
+      description 'Retrieves status of the resource for the actual user'
+      parameter name: :resource_id, in: :path, type: :string
+      produces 'application/json'
+      operationId 'getResourceStatus'
+
+      let(:resource_id) { create(:resource, name: 'ruby_course').id }
```
Comment (PabloAbara) [inline spec/requests/api/resources_controller_spec.rb:226]: Detail: ¿Why specify "name"? The factory could manage without input.

File: app/controllers/api/resources_controller.rb [hunk 5]
```diff
@@ -15,7 +17,10 @@ def index
 
     def average_evaluation
       resource = Resource.find(params[:resource_id])
-      evaluation = calculate_average_evaluation(resource.id)
+      evaluation =
+        Resources::ResourceEvaluationService.new(
```
Comment (PabloAbara) [inline app/controllers/api/resources_controller.rb:21]: ¿Why is this refactor on this PR?  It was for functional issues or needs?

File: app/models/completed_resource.rb [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+#  id          :bigint           not null, primary key
+#  resource_id :bigint           not null
+#  user_id     :bigint           not null
+#  created_at  :datetime         not null
+#  updated_at  :datetime         not null
+#
+class CompletedResource < ApplicationRecord
+  belongs_to :user
+  belongs_to :resource
+
```
Comment (PabloAbara) [inline app/models/completed_resource.rb:14]: I think we are adding model validation for "presence" of user and resource as well as validation table (that is OK!).

File: db/migrate/20221102182102_add_uniqueness_of_completed_resources.rb [hunk 7]
```diff
@@ -0,0 +1,7 @@
+class AddUniquenessOfCompletedResources < ActiveRecord::Migration[7.0]
+  def change
+    remove_index :completed_resources, :user_id
```
Comment (PabloAbara) [inline db/migrate/20221102182102_add_uniqueness_of_completed_resources.rb:3]: In the future, you could rollback migrations that where corrected within the PR. But this works fine.

**Review by PabloAbara — APPROVED** [review]
Just minor comments apart from functionality. It is approved.

File: spec/requests/api/resources_controller_spec.rb (hunk 4 above)
Comment (ristofer) [reply spec/requests/api/resources_controller_spec.rb:235]: It is true. I copied the spec of learning units.

**Review by valentinjadot — APPROVED** [review]
cool!

**Pull request merged**